internal/v5/auth.go

```diff
@@ -293,20 +297,44 @@ func areAllowedEntities(entityIds []*router.ResolvedURL, allowedEntities string)
 // AuthorizeEntity checks that the given HTTP request
 // can access the entity with the given id.
 func (h *ReqHandler) AuthorizeEntity(id *router.ResolvedURL, req *http.Request) error {
-	baseEntity, err := h.Cache.BaseEntity(id.UserOwnedURL(), charmstore.FieldSelector("acls", "developmentacls"))
+	baseEntity, err := h.Cache.BaseEntity(id.UserOwnedURL(), charmstore.FieldSelector("acls", "developmentacls", "stableacls"))
 	if err != nil {
 		if errgo.Cause(err) == params.ErrNotFound {
 			return errgo.WithCausef(nil, params.ErrNotFound, "entity %q not found", id)
 		}
 		return errgo.Notef(err, "cannot retrieve entity %q for authorization", id)
 	}
-	acls := baseEntity.ACLs
-	if id.Development {
-		acls = baseEntity.DevelopmentACLs
+	acls, err := h.getACLs(baseEntity, id)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
 	}
 	return h.authorizeWithPerms(req, acls.Read, acls.Write, id)
 }
 
+// getACLs returns the ACLs to use for the given base entity and resolved id.
+// What ACLs apply here? If a channel has been explicitly specified in the
+// URL, then, assuming the entity has been already resolved and fond for
+// that channel, channel specific ACLs apply. Otherwise, the URL has no
+// channels, and so stable ACLs apply if the entity has been published as
+// stable, then development ACLs if the entity is in development, otherwise
+// the system falls back to unpublished ACLs.
+func (h *ReqHandler) getACLs(baseEntity *mongodoc.BaseEntity, id *router.ResolvedURL) (mongodoc.ACL, error) {
```

Comment:
As written this has an interesting property that people can lose access when a charm is published. Say for example ACLs.Read contains {"tom", "dick"} and DevelopmentACLs.Read contains {"tom", "harry"} then if dick tries to install from the charm with no channel, then before it is published at all he will be able to deploy it, after it is publised to development he will not. Perhaps if the channel isn't specified the ACL should be the union of all ACLs for channels to which the charm is published.